### ISSUE TITLE
GRUB installation problem

### ISSUE THREAD

**Issue by macdems** [warewulf/warewulf3#16]
I have noticed that GRUB installation in `80-mkbootable`  is very badly organized. It is importable and very likely to fail on many systems. My tests (with Ubuntu) have shown that Warewulf does very well for stateless and statelite provisioning, however if fails with stateful one.

I would suggest to refactor this. Here are my suggestions:

1) `grub.cfg`  or `grub.conf` should probably be created per VNFS — it will be much easier to consider differences between different distros and will allow to setup stateful node with different kernel than set in the boostrap (can make sense if the provisioned system is different one than the master one).

2) `grub-install` command should be defined in the database per VNFS. This will make it safe for any future system changes and maybe will even allow to use different bootloader than grub. The safest thing is to define the whole command line with some template for node-specific data (like boot partition).

3) For stateful provisioning Warewulf should not try to call `/sbin/init`. Probably `wwgetvnfs` should not test for it. Instead after provisioning the system should be rebooted and set to booting from disk (for example provision `bootlocal` parameter should allow value `next:xxx` which means: boot wit warewulf, provision and set `bootlocal` to _xxx_.

I suggest to discuss this issue as the refactor would probably require many changes in many places.

**Comment by bensallen** [comment]
Overall 80-mkbootable has seen a lot of evolution across the last few years. It's largely tested on RHEL6 and 7.

1. It probably makes sense to allow a configurable override for specifying where to find a grub.cfg and leave the existing functionality as well. grub2-install is fairly standardized across distros now, so I'm not sure how much need there actually is for this.

2. Configurable override to run a given "WWBOOTINSTALLCMD" makes sense.

3. See provision/initramfs/capabilities/setup-filesystems/99-postreboot. Note, I'm pretty sure /tmp/stateful-root isn't being created anywhere, so this needs to be patched.

All the changes should be within 80-mkbootable. I'm making a few changes in #12 to 80-mkbootable, so it would be good to wait before doing any work until that work is merged.

**Comment by macdems** [comment]
In Ubuntu `/usr/sbin/grub2-install` is renamed to `/usr/sbin/grub2-install`.

**Comment by jmstover** [comment]
@bensallen Hrmm... if `/tmp/stateful-root` isn't being created, should add it into `provision/initramfs/capabilities/setup-filesystems/20-filesystems` at the beginning where it looks for a '/' mountpoint in FILESYSTEMS.

@macdems And yes, the stateful provisioning is a hack. It _mostly_ works with RHEL based systems, except for when it doesn't... Calling `/sbin/init` is the sanest choice IMO. We'd need to make a change to the PXE file *after* it PXE boots to provision, otherwise you're going to be caught in a loop. I mean, calling /sbin/init from the bootstrap isn't different than what the kernel does when it gets loaded from grub unless you're wanting to *boot* from a kernel that's different than what you provision from.

There's nothing existing to to callback to the provisioner and get the state of a node or act on some event.

WW itself isn't concerned with that, only that it gets an image. Once it's provisioned, you can set bootlocal on the node, and it'll just go to the next device on PXE booting, but that's after it has already been provisioned.

Up for other options here.

**Comment by bensallen** [comment]
@jmstover Was thinking of just removing the existing check for /tmp/stateful-root, and as long as there's something mounted at /newroot then reboot. If nothing is mounted at all to /newroot, exit 2 to fail.

Ideally we'd actually fail and bail out into a debug shell, so the user could troubleshoot further. Or... do we prefer to reboot when we fail?

**Comment by jmstover** [comment]
Keep normal logic. The stateless image is mounted at `/newroot` as well. Postreboot should be doing the right thing and sync'ing the disk... just to be sure.

But _just_ a reboot isn't going to do anything without changing the PXE file, or changing the boot order on the node...

**Comment by bensallen** [comment]
@macdems Ok lets start gathering a list of all the places grub-install and grub2-install can be located and what version of grub it actually is. I'd like to at least cover the common distros automatically.

Ubuntu 14.04 - /usr/sbin/grub-install - grub 2.02
RHEL6 - /sbin/grub-install - grub 0.97
RHEL7 - /usr/sbin/grub2-install - grub 2.02

Adding on to the list of things todo in mkbootable is EFI support, which is on my short list of tasks.

**Comment by bensallen** [comment]
@jmstover Right so why do we even have the reboot in postreboot then?

**Comment by jmstover** [comment]
That I don't know... @gmkurtzer might know why a reboot was put there. I mean, a reboot would make sense if you _always_ did something like....

```
# wwsh -y ipmi powercycle node01; sleep 30; wwsh -y provision set node01 --bootlocal=exit
```
And hope that provisioning was successful.

But in general.... no an actual reboot there doesn't make sense.

**Comment by macdems** [comment]
@jmstover 
> And yes, the stateful provisioning is a hack. It mostly works 

And this is why I think the current `80-mkbootable` is wrong in my opinion. Bootloader installation should be configurable with provision. Maybe something similar to the current behavior can be considered a sane default.

@bensallen 
> Ok lets start gathering a list of all the places grub-install and grub2-install can be located...

Looks fine, but no hard-coded list will always work. Instead I would suggest to allow for multiple `/etc/warewulf/defaults/provision` configs (like multiple `wwvnfs` configs) I the user can define everything required to provision any machine (VNFS, bootloader installation, etc.)

Probably it may be useful to prepare some scripts similar to `wwmkchroot` that will setup chroot, create VNFS and set the filesystems bootloader provisioning settings for the supported distros.

@jmstover 
> But in general.... no an actual reboot there doesn't make sense.

I does in many IPMI systems. It is not uncommon to force just one node boot to PXE and afterwards the node returns back to the default HD boot. If available, this is much better approach than changing `provision --bootlocal`. Or even setting the node boot order to HD, PXE makes sense as in the first boot there will be no bootloader on an unformatted HD, so the machine will automatically start from the net and gets provisioned.

**Comment by jmstover** [comment]
@macdems 
> It is not uncommon to force just one node boot to PXE and afterwards 
> the node returns back to the default HD boot.

And this is opposite the main use case of Warewulf where you're PXE booting as the primary always. The provisioner controls what the node does on boot through the PXE file.

Stateful provisioning was an afterthought, and added on when requested. Its never been the greatest, but it's mostly functional in getting parts of the image to disk. At work, we use "stateful" in that we provision the image to disk (just so as much RAM is free as possible), but we _always_ PXE boot and drop a new image down on reboot... even login nodes. Special case nodes get set bootlocal. Usually nodes that get provisioned, then segmented off the main network...

So, Warewulf was designed as a stateless provisioner, and the disk stuff was added on as a "nice to have" feature, but it wasn't designed with writing to disk in mind, so the support is... highly lacking.

**Comment by macdems** [comment]
@jmstover The use case you describe is what I call state-lite provisioning. This and full stateless works very well with Warewulf! This is actually the very reason I have decided to take a closer look at the project and I am implementing it in my clusters.

However, in my opinion, stateful is a nice-to-have feature, but it must be done right. Hence, I suggest keeping this discussion open and postponing this feature for 3.8 (after #12 and #17 are closed). Next week I will try to do a quick hack to `80-mkbootable` so it supports Ubuntu similarly to the current approach for RHEL and submit a pull request. And then we can think on more portable distro-agnostic solution for the future.

**Comment by jmstover** [comment]
If I remember my hack to mkbootable, what will probably want to be done is:

Check for the existence of `$NEWROOT/etc/redhat-release` or `$NEWROOT/etc/debian_version` and set a `GRUBCMD` variable based on the distro.

Change the grub-installer calls to be `${GRUBCMD}` instead... The logic should hopefully still work out to be the same. 

If you aren't changing much with the kernels, you _shouldn't_ need to modify the grub config then... So, hopefully as a first pass, that should cover more use cases.

**Comment by bensallen** [comment]
Closing this with #24 merged.
